apps/els_lsp/src/els_parser.erl

```diff
@@ -66,46 +79,97 @@ parse_forms(Forms) ->
 
 -spec parse_form(erlfmt_parse:abstract_node()) -> deep_list(poi()).
 parse_form({raw_string, Anno, Text}) ->
-  Start = erlfmt_scan:get_anno(location, Anno),
-  {ok, RangeTokens, _EndLocation} = erl_scan:string(Text, Start, [text]),
-  find_attribute_tokens(RangeTokens);
+  StartLoc = erlfmt_scan:get_anno(location, Anno),
+  RangeTokens = scan_text(Text, StartLoc),
+  case parse_incomplete_tokens(RangeTokens) of
+    {ok, Form} ->
+      parse_form(Form);
+    error ->
+      find_attribute_tokens(RangeTokens)
+  end;
 parse_form(Form) ->
   Tree = els_erlfmt_ast:erlfmt_to_st(Form),
   POIs = points_of_interest(Tree),
   POIs.
 
+-spec scan_text(string(), {erl_anno:line(), erl_anno:column()})
+               -> [erlfmt_scan:token()].
+scan_text(Text, StartLoc) ->
+  PaddedText = pad_text(Text, StartLoc),
+  {ok, Tokens, _Comments, _Cont} = erlfmt_scan:string_node(PaddedText),
+  ensure_dot(Tokens).
+
+-spec parse_incomplete_tokens([erlfmt_scan:token()])
+                             -> {ok, erlfmt_parse:abstract_node()} | error.
+parse_incomplete_tokens([{dot, _}]) ->
+  error;
+parse_incomplete_tokens(Tokens) ->
+  case erlfmt_parse:parse_node(Tokens) of
+    {ok, Form} ->
+      {ok, Form};
+    {error, {ErrorLoc, erlfmt_parse, _Reason}} ->
+      TrimmedTokens = tokens_until(Tokens, ErrorLoc),
+      parse_incomplete_tokens(TrimmedTokens)
+  end.
+
+%% @doc Drop tokens after given location but keep final dot, to preserve its
+%% location
+-spec tokens_until([erlfmt_scan:token()], erl_anno:location())
+                  -> [erlfmt_scan:token()].
+tokens_until([_Hd, {dot, _} = Dot], _Loc) ->
```

Comment:
Are we sure this cannot be invoked with `[]`?

Reply:
dialyzer was sure :). it's because ensure_dot will only return a non empty list of tokens

Reply:
btw I did not test cases which cannot even be scanned, the error handling might be lacking in those cases. will look at that.

Reply:
looked at scanning errors:
if text cannot be tokenized then `erlfmt:read_nodes_string/2` returns an error (the error from `erl_scan`) (no forms at all). The new code only handles `raw_string` quasi-forms, which are only returned by `erlfmt` when scanning was successful but parsing was not for some forms. So in this part of the code it is not necessary to handle text which cannot be tokenized.

It would be possible to trim the text at the scanning error position, but that's a different code path and Im not sure how useful it is
- a scanning error because of a not yet closed quote (string or atom): no completion needed within a quote
- other scanning error which is not incomplete code like bad integer `16#g`: compiler diagnostics warns about these, probably best to expect the user to fix these before proceeding (and expecting ELS to work properly) this is more a tree-sitter type of work to fix heuristically.

Reply:
To me the current approach seems good enough for now. Should we look at these possibilities as potential follow-ups?

Reply:
One thing we could do is to add a testcase explaining the limitations of the current approach in that specific situation and leave a commented out version of what we would like instead.

Reply:
added a testcase for scanning errors and commented future opportunities